actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if not @isProtocolSupported()
+        @close()
+      else
+        ActionCable.log("WebSocket using '" + @protocol() + "' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
 
-    close: ->
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+      if not @isProtocolSupported()
+        ActionCable.log("WebSocket is permanently closing because there is no common subprotocol. Client supported protocols: " + protocols.join(','))
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+      else
+        @disconnect(new ActionCable.Events.Disconnected(event.code, event.reason, true))
 
     error: ->
       ActionCable.log("WebSocket onerror event")
-      @disconnect()
```

Comment:
https://www.w3.org/TR/websockets/#concept-websocket-close-fail

> If the user agent was required to fail the WebSocket connection or the WebSocket connection is closed with prejudice, fire a simple event named error at the WebSocket object.

Firefox and Safari do not consistently fire this `error` event at all, but they all seem to fire the `close` event consistently.